libdnf/rpm/reldep_list.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright (C) 2018-2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/rpm/reldep_list.hpp"
+#include "libdnf/rpm/reldep.hpp"
+#include "reldep_list_impl.hpp"
+#include "sack_impl.hpp"
+#include "solv/reldep_parser.hpp"
+
+// libsolv
+extern "C" {
+#include <solv/dataiterator.h>
+#include <solv/queue.h>
+}
+
+
+namespace libdnf::rpm {
+
+ReldepList::ReldepList(const ReldepList & src)
+        : pImpl(new Impl(*src.pImpl))
+{}
+
+ReldepList::ReldepList(ReldepList && src) noexcept
+        : pImpl(std::move(src.pImpl))
+{}
+
+ReldepList::ReldepList(Sack * sack)
+        : pImpl(new Impl(sack))
+{}
+
+ReldepList::~ReldepList() = default;
+
+ReldepId ReldepList::get_id(int index) const noexcept
+{
+    return ReldepId(pImpl->queue[index]);
+}
+
+ReldepList & ReldepList::operator=(ReldepList && src) noexcept
+{
+    pImpl.swap(src.pImpl);
+    return *this;
+}
+
+bool ReldepList::operator!=(const ReldepList & other) const noexcept { return !(*this == other); }
+bool ReldepList::operator==(const ReldepList & other) const noexcept
+{
+    auto & this_queue = pImpl->queue;
+    auto & other_queue = other.pImpl->queue;
+    auto this_count = this_queue.size();
+    if (this_count != other_queue.size())
+        return false;
+
+    for (int i = 0; i < this_count; i++) {
+        if (this_queue[i] != other_queue[i]) {
+            return false;
+        }
+    }
+
+    return pImpl->sack->pImpl->pool == other.pImpl->sack->pImpl->pool;
```

Comment:
OK. The code compare two queues. And after for cycle pass it is found that the `pools` are different. Is the order correct/optimal?

Reply:
I am also not sure